storage/innobase/handler/handler0alter.cc

```diff
@@ -654,9 +654,13 @@ static MY_ATTRIBUTE((warn_unused_result)) bool innobase_need_rebuild(
   Alter_inplace_info::HA_ALTER_FLAGS alter_inplace_flags =
       ha_alter_info->handler_flags & ~(INNOBASE_INPLACE_IGNORE);
 
-  if (Encryption::none_explicitly_specified(
-          ha_alter_info->create_info->explicit_encryption,
-          ha_alter_info->create_info->encrypt_type.str) ||
+  if ((Encryption::none_explicitly_specified(
+           ha_alter_info->create_info->explicit_encryption,
+           ha_alter_info->create_info->encrypt_type.str) &&
+       ((old_table->s->tablespace == nullptr && !old_table->s->system) ||
+        (old_table->s->tablespace != nullptr &&
+         strcmp(old_table->s->tablespace, dict_sys_t::s_file_per_table_name) ==
```

Comment:
Please find a simpler way. This is error prone. Also please add a comment why this code says redbuild required when encryption='n'.

Reply:
Try this : dd::is_tablespace_encrypted() 
THis is available everywhere.. if not just include relevant header.

Reply:
The name is misleading. It has out parameter which tells if tablespace is general or not.

Reply:
dd::is_tablespace_encrypted works with dd::Table, I have Table here. I have reused tablespace_is_shared_space.